de.fhdo.lemma.service.open_api/src/de/fhdo/lemma/service/open_api/LemmaDataSubGenerator.xtend

```diff
@@ -0,0 +1,197 @@
+            c.complexTypes.add(newDataStructure)
+            addDataFieldsToDataStructure(newDataStructure, name, schema)
+            createdDataStructures.put(
+                newDataStructure.buildQualifiedName(seperator), newDataStructure
+            )
+            return newDataStructure
+        } else {
+            return foundObject
+        }
+    }
+
+    def ListType getOrCreateListStructure(Context c, String name, Schema schema) {
+        var foundObject = findListStructure(c, name)
+        if (foundObject === null) {
+            val newListStructure = DATA_FACTORY.createListType
+            newListStructure.name = '''«StringUtils.capitalize(name)»List'''
+            c.complexTypes.add(newListStructure)
+            addDataFieldsToListStructure(newListStructure, name, schema)
+            createdListStructures.put(
+                newListStructure.buildQualifiedName(seperator), newListStructure
+            )
+            return newListStructure
+        } else {
+            return foundObject
+        }
+    }
+
+    def void addDataFieldsToDataStructure(
+        DataStructure ds, String structureName, Schema structureSchema
+    ) {
+        if (structureSchema.properties !== null) {
+            structureSchema.properties.forEach[
+                name, schema|addDataFieldsToDataStructure(ds, name, schema)
+            ]
+        } else {
+            val newDataField = generateDataField(structureName, structureSchema)
+            ds.dataFields.add(newDataField)
+        }
+    }
+
+    def addDataFieldsToListStructure(
+        ListType ds, String structureName, Schema structureSchema
+    ) {
+        val newDataField = generateDataField(structureName, structureSchema)
+        ds.dataFields.add(newDataField)
+    }
+
+    def DataField generateDataField(
+        String structureName, Schema structureSchema
+    ) {
+        val newDataField = DATA_FACTORY.createDataField
+        newDataField.name = StringUtils.lowerCase(structureName)
+
+        switch (structureSchema.type) {
+            case "integer": {
+                newDataField.primitiveType = OpenApiUtil.deriveIntType(structureSchema.format)
+            }
+            case "number": {
+                newDataField.primitiveType = OpenApiUtil.deriveNumberType(structureSchema.format)
+            }
+            case "string": {
+                newDataField.primitiveType = OpenApiUtil.deriveStringType(structureSchema.format)
+            }
+            case "boolean": {
+                newDataField.primitiveType = DATA_FACTORY.createPrimitiveBoolean
+            }
+            case "array": {
+                val itemSchema = structureSchema as ArraySchema
+                val listStructure = getOrCreateListStructure(
+                    myContext, structureName, itemSchema.items
+                )
+                newDataField.complexType = listStructure
+            }
+            case null: {
```

Comment:
It would be way more readable and comprehensible to break this method into two:
- The first method handles actual types (all previous cases).
- The second method handles the reference case (this case).